File: creational/typescript/factory_method_structured/creators.ts
import { Application, Document, PaymentProcessor } from './interfaces';
import { TextDocument, PDFDocument, HtmlDocument, CreditCardProcessor, PayPalProcessor, CryptoProcessor } from './concrete_products';

export class TextEditor extends Application {
  protected createDocument(): Document { return new TextDocument(); }
}

export class PDFViewer extends Application {
  protected createDocument(): Document { return new PDFDocument(); }
}

export class WebBrowser extends Application {
  protected createDocument(): Document { return new HtmlDocument(); }
}

export enum DocType { text, pdf, html }

export class DocumentFactory {
  static createDocument(type: DocType): Document {
    switch (type) {
      case DocType.text: return new TextDocument();
      case DocType.pdf: return new PDFDocument();
      case DocType.html: return new HtmlDocument();
    }
  }
}

export enum PaymentMethod { creditCard, payPal, crypto }

export class PaymentFactory {
  static createProcessor(method: PaymentMethod): PaymentProcessor {
    switch (method) {
      case PaymentMethod.creditCard: return new CreditCardProcessor();
      case PaymentMethod.payPal: return new PayPalProcessor();
      case PaymentMethod.crypto: return new CryptoProcessor();
    }
  }
}

File: creational/dart/factory_method_structured/creators.dart
import 'interfaces.dart';
import 'concrete_products.dart';

class TextEditor extends DocumentApplication {
  @override
  Document createDocument() {
    return TextDocument();
  }
}

class PDFViewer extends DocumentApplication {
  @override
  Document createDocument() {
    return PDFDocument();
  }
}

class WebBrowser extends DocumentApplication {
  @override
  Document createDocument() {
    return HtmlDocument();
  }
}

enum DocType { text, pdf, html }

class DocumentFactory {
  static Document createDocument(DocType type) {
    switch (type) {
      case DocType.text:
        return TextDocument();
      case DocType.pdf:
        return PDFDocument();
      case DocType.html:
        return HtmlDocument();
    }
  }
}

enum PaymentMethod { creditCard, payPal, crypto }

class PaymentFactory {
  static PaymentProcessor createProcessor(PaymentMethod method) {
    switch (method) {
      case PaymentMethod.creditCard:
        return CreditCardProcessor();
      case PaymentMethod.payPal:
        return PayPalProcessor();
      case PaymentMethod.crypto:
        return CryptoProcessor();
    }
  }
}
